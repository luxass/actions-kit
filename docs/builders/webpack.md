# Webpack Builder

Build GitHub Actions using Webpack.

## Installation

::: code-group

```bash [pnpm]
pnpm add -D actions-kit @actions-sdk/webpack-builder
```

```bash [yarn]
yarn add -D actions-kit @actions-sdk/webpack-builder
```

```bash [npm]
npm install -D actions-kit @actions-sdk/webpack-builder
```

```bash [bun]
bun install -D actions-kit @actions-sdk/webpack-builder
```

:::

Set the `builder` in your `actions-kit.config.ts` file:

```ts [actions-kit.config.ts]
import webpack from "@actions-sdk/webpack-builder";
import { defineConfig } from "actions-kit/config";

export default defineConfig({
  builder: webpack({
    ...webpackOptions,
  }),
});
```

### Customizing Webpack

You can customize the Webpack options by passing the options directly into the `webpack` function.

```ts [actions-kit.config.ts]
import webpack from "@actions-sdk/webpack-builder";
import { defineConfig } from "actions-kit/config";

export default defineConfig({
  builder: webpack({
    ...webpackOptions,
  }),
});
```
