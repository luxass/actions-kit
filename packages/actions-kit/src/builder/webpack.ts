// TODO: remove this at some point
//       will need to debug why the dev build fails
//       when its not there.
import "zod";
import type { Configuration as WebpackConfig, Stats } from "webpack";
import { join, resolve } from "node:path";
import { defu } from "defu";
import actionsKit from "unplugin-actions-kit/webpack";
import type { Config } from "../config";
import { inferModuleType, inferOutputFilename } from "../utils";

export async function build(config: Config) {
	const webpack = await import("webpack").then((m) => m.default);
	const outputFileName = await inferOutputFilename(config);
	const libraryType = await inferModuleType(config, outputFileName);

	const webpackConfig = config.webpack;
	const webpackOptions = defu(webpackConfig, {
		target: "node",
		mode: "production",
		entry: "./src/index.ts",
		output: {
			path: resolve(process.cwd(), "dist"),
			filename: outputFileName,
			library: {
				type: libraryType === "esm" ? "module" : "commonjs2",
			},
		},
		resolve: {
			extensions: [".ts", ".js"],
		},
		optimization: {
			minimize: false,
		},
		devtool: false,
		module: {
			rules: [
				{
					test: /\.ts$/,
					exclude: [/node_modules/],
					loader: "builtin:swc-loader",
					options: {
						jsc: {
							parser: {
								syntax: "typescript",
							},
						},
					},
					type: "javascript/auto",
				},
			],
		},
		plugins: [
			actionsKit({
				// TODO: allow users to specify it.
				actionPath: join(process.cwd(), "./action.yml"),
				inject: true,
			}),
		],
		externals: {
			keytar: "commonjs keytar",
		},
	} satisfies WebpackConfig);

	const compiler = webpack.webpack(webpackOptions);

	const stats = await new Promise<Stats | undefined>((resolve, reject) =>
		compiler.run((err, stats) => {
			if (err) {
				reject(err);
			} else {
				resolve(stats as Stats);
			}
		}),
	);

	if (!stats) {
		throw new Error("could not build");
	}

	stats.toString({
		chunks: true,
		colors: true,
	});
}
