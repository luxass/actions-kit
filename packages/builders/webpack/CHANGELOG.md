# @actions-sdk/webpack-builder

## 0.2.6

### Patch Changes

- [#32](https://github.com/luxass/actions-kit/pull/32) [`cb0ba07`](https://github.com/luxass/actions-kit/commit/cb0ba0748e6aaf27049c152c8f7d562a0c01f54e) Thanks [@luxass](https://github.com/luxass)! - output correct path in build output

- Updated dependencies [[`d13ac04`](https://github.com/luxass/actions-kit/commit/d13ac043b044c46fea1f9e4d5c56ddb8543a8ed6)]:
  - actions-kit@0.2.7

## 0.2.5

### Patch Changes

- Updated dependencies []:
  - actions-kit@0.2.6

## 0.2.4

### Patch Changes

- [#28](https://github.com/luxass/actions-kit/pull/28) [`ed50302`](https://github.com/luxass/actions-kit/commit/ed503020a2181f1fe2e5bc26aae50e2adc8e411e) Thanks [@luxass](https://github.com/luxass)! - use new `inferOutput` builder utility to correctly set the outdir

- Updated dependencies [[`ed50302`](https://github.com/luxass/actions-kit/commit/ed503020a2181f1fe2e5bc26aae50e2adc8e411e)]:
  - actions-kit@0.2.5

## 0.2.3

### Patch Changes

- Updated dependencies [[`ae1a6ec`](https://github.com/luxass/actions-kit/commit/ae1a6ec04ad71f0adda06439877d65182719ba9a)]:
  - unplugin-actions-kit@0.2.3
  - actions-kit@0.2.4

## 0.2.2

### Patch Changes

- [`efedd0c`](https://github.com/luxass/actions-kit/commit/efedd0cf1448c1d480cddde2ef43a3939b325be6) Thanks [@luxass](https://github.com/luxass)! - run build before publish

- Updated dependencies [[`efedd0c`](https://github.com/luxass/actions-kit/commit/efedd0cf1448c1d480cddde2ef43a3939b325be6)]:
  - actions-kit@0.2.3
  - unplugin-actions-kit@0.2.2

## 0.2.1

### Patch Changes

- [#23](https://github.com/luxass/actions-kit/pull/23) [`46032ce`](https://github.com/luxass/actions-kit/commit/46032ce171655d8a2eb7277238664aa386485b3b) Thanks [@luxass](https://github.com/luxass)! - fix package exports and add missing cli entrypoint for `actions-kit`

- Updated dependencies [[`46032ce`](https://github.com/luxass/actions-kit/commit/46032ce171655d8a2eb7277238664aa386485b3b)]:
  - actions-kit@0.2.2

## 0.2.0

### Minor Changes

- [`bcf5300`](https://github.com/luxass/actions-kit/commit/bcf53008a191840aee634e040f4d9c53cfbf2a8b) Thanks [@luxass](https://github.com/luxass)! - change how builders work, for more information https://actions-kit.pages.dev/builders/
