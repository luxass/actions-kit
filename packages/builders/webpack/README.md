# @actions-kit/webpack-builder

Webpack builder for Actions Kit.
