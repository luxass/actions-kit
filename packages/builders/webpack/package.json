{
  "name": "@actions-sdk/webpack-builder",
  "version": "0.3.1",
  "description": "Webpack builder for Actions Kit",
  "type": "module",
  "author": {
    "name": "Lucas Nørgård",
    "email": "lucasnrgaard@gmail.com",
    "url": "https://luxass.dev"
  },
  "packageManager": "pnpm@10.15.1",
  "license": "MIT",
  "homepage": "https://github.com/luxass/actions-kit",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/luxass/actions-kit.git",
    "directory": "packages/builders/webpack"
  },
  "bugs": {
    "url": "https://github.com/luxass/actions-kit/issues"
  },
  "keywords": [
    "webpack",
    "actions-kit",
    "github actions",
    "actions"
  ],
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "types": "./dist/index.d.cts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsdown --tsconfig=./tsconfig.build.json",
    "dev": "tsdown --watch",
    "clean": "git clean -xdf dist node_modules",
    "lint": "eslint .",
    "typecheck": "tsc --noEmit"
  },
  "peerDependencies": {
    "actions-kit": "workspace:*",
    "webpack": "catalog:builders"
  },
  "dependencies": {
    "@swc/core": "catalog:builders",
    "defu": "catalog:prod",
    "swc-loader": "catalog:builders",
    "unplugin-actions-kit": "workspace:*"
  },
  "devDependencies": {
    "@actions-kit/tsconfig": "workspace:*",
    "@actions-kit/tsdown-config": "workspace:*",
    "@luxass/eslint-config": "catalog:linting",
    "@types/node": "catalog:types",
    "actions-kit": "workspace:*",
    "eslint": "catalog:linting",
    "publint": "catalog:dev",
    "tsdown": "catalog:dev",
    "typescript": "catalog:dev",
    "vitest": "catalog:testing",
    "webpack": "catalog:builders"
  }
}
