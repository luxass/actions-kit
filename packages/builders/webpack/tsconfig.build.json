{
  "extends": ["../../../tsconfig.base.build.json"],
  "include": ["webpack.d.ts", "src/**/*.ts"]
}
