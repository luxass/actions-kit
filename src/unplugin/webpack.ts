/**
 * This entry file is for webpack plugin.
 *
 * @module
 */

import { createWebpackPlugin } from "unplugin";
import type unplugin from "./";
import { unpluginFactory } from "./";

/**
 * Webpack plugin
 *
 * @example
 * ```ts
 * // webpack.config.js
 * module.exports = {
 *  plugins: [require("actions-kit/unplugin/webpack")()],
 * }
 * ```
 */
export default createWebpackPlugin(unpluginFactory) as typeof unplugin.webpack;
