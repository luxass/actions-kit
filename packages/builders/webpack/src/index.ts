import type { BuildOutput } from "actions-kit/builder";
import type { Configuration, Stats } from "webpack";
import { join } from "node:path";
import { defineBuilder } from "actions-kit/builder";
import { inferModuleType, inferOutput } from "actions-kit/builder-utils";
import { defu } from "defu";
import WebpackActionsKit from "unplugin-actions-kit/webpack";
import { webpack } from "webpack";

export default function webpackBuilder(options: Configuration = {}) {
  return defineBuilder({
    name: "webpack",
    build: async ({ cwd, config }) => {
      const { filename, dir } = await inferOutput(config);
      const libraryType = await inferModuleType(config, filename);

      const webpackOptions = defu(options, {
        target: "node",
        mode: "production",
        entry: "./src/index.ts",
        output: {
          path: dir,
          filename,
          library: {
            type: libraryType === "esm" ? "module" : "commonjs2",
          },
        },
        resolve: {
          extensions: [".ts", ".js"],
        },
        optimization: {
          minimize: false,
        },
        devtool: false,
        module: {
          rules: [
            {
              test: /\.m?js$/,
              exclude: /(node_modules)/,
              use: {
                loader: "swc-loader",
              },
            },
          ],
        },
        plugins: [
          WebpackActionsKit({
            // TODO: allow users to specify it.
            actionPath: join(cwd, "./action.yml"),
            inject: config.inject,
            autocomplete: config.autocomplete,
          }),
        ],
      } satisfies Configuration);

      const compiler = webpack(webpackOptions);

      const stats = await new Promise<Stats | undefined>((resolve, reject) => {
        if (compiler == null) {
          return reject(new Error("webpack compiler is not defined"));
        }

        compiler.run((err, stats) => {
          if (err) {
            reject(err);
          } else {
            resolve(stats as unknown as Stats);
          }
        })
      });

      if (!stats) {
        throw new Error("could not build");
      }

      const json = stats.toJson({
        all: false,
        assets: true,
      });

      const output: BuildOutput[] = [];

      if (json.assets == null) {
        throw new Error("could not gather bundle information");
      }

      const assets = json.assets;

      for (const asset of assets) {
        output.push({
          name: asset.name,
          path: join(dir, asset.name),
          size: asset.size,
        });
      }

      return output;
    },
  });
}
