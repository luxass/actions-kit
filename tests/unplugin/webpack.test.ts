import { readFile } from "node:fs/promises";
import path, { join } from "node:path";
import { expect, it } from "vitest";
import { fromFileSystem, testdir } from "vitest-testdirs";
import { type Configuration, webpack as createWebpack, type Stats } from "webpack";
import ActionKitPlugin from "../../src/unplugin/webpack";

async function webpack(testdirPath: string, config: Configuration): Promise<Stats | undefined> {
  return new Promise((resolve, reject) => {
    const compiler = createWebpack({
      optimization: {
        minimize: false,
      },
      output: {
        path: path.resolve(testdirPath),
        filename: `${Date.now()}-bundle.js`,
      },
      module: {
        rules: [
          {
            test: /\.tsx?$/,
            loader: "ts-loader",
            exclude: /node_modules/,
            options: {
              transpileOnly: true,
              experimentalWatchApi: true,
            },
          },
          ...(config.module?.rules || []),
        ],
        ...config.module,
      },
      resolve: {
        preferRelative: true,
        extensions: [".tsx", ".ts", ".js"],
        extensionAlias: {
          ".js": [".js", ".ts"],
          ".cjs": [".cjs", ".cts"],
          ".mjs": [".mjs", ".mts"],
        },
      },
      target: "node",
      mode: "production",
      externals: ["@actions/core"],
      externalsType: "commonjs",
      ...config,
    });

    compiler.run((err, stats) => {
      if (err) {
        reject(err);
      }

      resolve(stats);
    });
  });
}

it("expect `actions-kit.d.ts` to be generated", async () => {
  const directoryJson = await fromFileSystem("./tests/fixtures/action.yaml");
  const testdirPath = await testdir(directoryJson);

  expect(testdirPath).toBeDefined();

  const result = await webpack(testdirPath, {
    entry: join(testdirPath, "index.ts"),
    plugins: [
      ActionKitPlugin({
        actionPath: join(testdirPath, "action.yaml"),
      }),
    ],
  });

  const json = result?.toJson();
  expect(json).toBeDefined();

  expect(json?.errors).toBeDefined();

  if ((json?.errors || []).length > 0) {
    console.error(json?.errors);
  }

  expect(json?.errors).toHaveLength(0);

  const file = json!.assetsByChunkName!.main;
  const content = await readFile(path.join(testdirPath, file![0]!), "utf-8");

  expect(content).toMatchSnapshot();

  const dtsOutput = await readFile(join(testdirPath, "actions-kit.d.ts"), "utf-8");

  expect(dtsOutput).toMatchSnapshot();
});

it("expect `actions-kit.d.ts` to include `ACTION_INPUTS`", async () => {
  const directoryJson = await fromFileSystem("./tests/fixtures/action.yaml");
  const testdirPath = await testdir(directoryJson);

  expect(testdirPath).toBeDefined();

  const result = await webpack(testdirPath, {
    entry: join(testdirPath, "index.ts"),
    plugins: [
      ActionKitPlugin({
        actionPath: join(testdirPath, "action.yaml"),
        injectInputs: true,
      }),
    ],
  });

  const json = result?.toJson();
  expect(json).toBeDefined();

  expect(json?.errors).toBeDefined();

  if ((json?.errors || []).length > 0) {
    console.error(json?.errors);
  }

  expect(json?.errors).toHaveLength(0);

  const file = json!.assetsByChunkName!.main;
  const content = await readFile(path.join(testdirPath, file![0]!), "utf-8");

  const dtsOutput = await readFile(join(testdirPath, "actions-kit.d.ts"), "utf-8");

  expect(dtsOutput).toMatchSnapshot();
  expect(content).toMatchSnapshot();

  expect(content).toContain("ACTION_INPUTS");
  expect(dtsOutput).toContain("ACTION_INPUTS");
});
