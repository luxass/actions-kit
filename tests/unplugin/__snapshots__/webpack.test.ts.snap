// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expect \`actions-kit.d.ts\` to be generated 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	"use strict";

;// external "@actions/core"
const core_namespaceObject = require("@actions/core");
;// ./.vitest-testdirs/vitest-webpack-expect-actions-kit-d-ts-to-be-generated/index.ts

async function run() {
    const type = core_namespaceObject.getInput("type");
    const whoToGreet = core_namespaceObject.getInput("who-to-greet");
    const message = \`Hello \${whoToGreet}!\`;
    core_namespaceObject.info(message);
    if (type === "error") {
        core_namespaceObject.setFailed("This is an error!");
    }
    core_namespaceObject.setOutput("message", message);
}
run().catch((err) => {
    console.error(err);
    core_namespaceObject.setFailed(err.message);
});

/******/ })()
;"
`;

exports[`expect \`actions-kit.d.ts\` to be generated 2`] = `
"// generated by 'actions-kit'
        import "@actions/core";

        

declare module "@actions/core" {

  type InputNames = "type" | "who-to-greet";

  export function getInput(name: InputNames, options?: InputOptions): string;
}
"
`;

exports[`expect \`actions-kit.d.ts\` to include \`ACTION_INPUTS\` 1`] = `
"// generated by 'actions-kit'
        import "@actions/core";

        
declare global {
  export const ACTION_INPUTS: {
  "type": "type",
  "who-to-greet": "who-to-greet"
};
}

declare module "@actions/core" {

  type InputNames = "type" | "who-to-greet";

  export function getInput(name: InputNames, options?: InputOptions): string;
}
"
`;

exports[`expect \`actions-kit.d.ts\` to include \`ACTION_INPUTS\` 2`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	"use strict";

;// external "@actions/core"
const core_namespaceObject = require("@actions/core");
;// ./.vitest-testdirs/vitest-webpack-expect-actions-kit-d-ts-to-include-ACTION_INPUTS/index.ts
globalThis.ACTION_INPUTS = {
    "type": "type",
    "who-to-greet": "who-to-greet"
};

async function run() {
    const type = core_namespaceObject.getInput("type");
    const whoToGreet = core_namespaceObject.getInput("who-to-greet");
    const message = \`Hello \${whoToGreet}!\`;
    core_namespaceObject.info(message);
    if (type === "error") {
        core_namespaceObject.setFailed("This is an error!");
    }
    core_namespaceObject.setOutput("message", message);
}
run().catch((err) => {
    console.error(err);
    core_namespaceObject.setFailed(err.message);
});
;

/******/ })()
;"
`;
